File: LaneAdviser.py
import copy
import random
import config as cfg
from random import randrange
import json
import math
import numpy
#import myGraphic

from get_inter_length_info import Data

inter_length_data = Data()

RESOLUTION = 2  # According to gen_advise.cpp

class LaneAdviser:
    def __init__(self):
            # Number of the boxes in one dimention
            num_di = RESOLUTION*(LANE_NUM_PER_DIRECTION*2+1)

            # 2D matrix of timing for advising
            timeMatrix = [[0]* num_di for i in range(num_di)]

            # Load the information of which boxes a trajectory will affect
            with open('advise_info/advise_info'+str(LANE_NUM_PER_DIRECTION)+'.json', 'r') as file:
                lane_dict = json.load(file)

            # Record the lane advice on each directions && initialize
            all_default_lane = dict()
            for direction in range(4):
                all_default_lane[(direction, 'L')] = direction*LANE_NUM_PER_DIRECTION
                all_default_lane[(direction, 'R')] = direction*LANE_NUM_PER_DIRECTION + (LANE_NUM_PER_DIRECTION -1)
                all_default_lane[(direction, 'S')] = (all_default_lane[(direction, 'L')] + all_default_lane[(direction, 'R')])//2

            count_advised_not_secheduled_car_num = dict()
            for lane_id in range(4*LANE_NUM_PER_DIRECTION):
                count_advised_not_secheduled_car_num[(lane_id, 'L')] = 0
                count_advised_not_secheduled_car_num[(lane_id, 'R')] = 0
                count_advised_not_secheduled_car_num[(lane_id, 'S')] = 0

    def copyMatrix(self):
        return copy.deepcopy(timeMatrix)

    # Update the table (final version for simulation usage)
    def updateTableFromCars(self, n_sched_car, advised_n_sched_car):
        resetTable()

        for car in n_sched_car:
            updateTable(car.lane, car.current_turn, car.OT+car.D, timeMatrix)

        # Case 1: update with the latest car
        '''
        for car in advised_n_sched_car:
            updateTable(car.lane, car.current_turn, car.position/ V_MAX, timeMatrix)
        #'''
        # Case 2: add costs
        '''
        for car in advised_n_sched_car:
            updateTableAfterAdvise(car.lane, car.current_turn, car.length, timeMatrix)
        #'''

        # Case 3: (case 1 + considering the halt) update with the latest car
        '''
        latest_delay_list = dict()
        cars_on_lanes = dict()
        for idx in range(4*LANE_NUM_PER_DIRECTION):
            cars_on_lanes[idx] = []
        for car in n_sched_car:
            cars_on_lanes[car.lane].append(car)
        for idx in range(4*LANE_NUM_PER_DIRECTION):
            sorted_car_list = sorted(cars_on_lanes[idx], key=lambda car: car.position, reverse=true)
            if len(sorted_car_list) > 0:
                latest_delay_list[idx] = sorted_car_list[0].D
            else:
                latest_delay_list[idx] = 0

        # Consider the halting time
        for car in advised_n_sched_car:
            if car.H_should_halt == true:
                updateTable(car.lane, car.current_turn, car.position/ V_MAX+latest_delay_list[idx], timeMatrix)
            else:
                updateTable(car.lane, car.current_turn, car.position/ V_MAX, timeMatrix)
        #'''

        # Case 4: (case 1 + considering the halt) update with the latest car  "Using desired_lane!!!!!!!""

        latest_delay_list = dict()
        cars_on_lanes = dict()
        for idx in range(4*LANE_NUM_PER_DIRECTION):
            cars_on_lanes[idx] = []
        for car in n_sched_car:
            cars_on_lanes[car.desired_lane].append(car)
        for idx in range(4*LANE_NUM_PER_DIRECTION):
            sorted_car_list = sorted(cars_on_lanes[idx], key=lambda car: car.position, reverse=true)
            if len(sorted_car_list) > 0:
                latest_delay_list[idx] = sorted_car_list[0].D
            else:
                latest_delay_list[idx] = 0

        for car in advised_n_sched_car:
            if (car.CC_state != None) &&  ("Platoon" in car.CC_state):
                updateTable(car.lane, car.current_turn, car.position/ V_MAX+latest_delay_list[car.desired_lane], timeMatrix)
            else:
                updateTable(car.lane, car.current_turn, car.position/ V_MAX, timeMatrix)


        # Count car number on each lane of advised but not scheduled
        for lane_id in range(4*LANE_NUM_PER_DIRECTION):
            count_advised_not_secheduled_car_num[(lane_id, 'L')] = 0
            count_advised_not_secheduled_car_num[(lane_id, 'R')] = 0
            count_advised_not_secheduled_car_num[(lane_id, 'S')] = 0
        for car in advised_n_sched_car:
            count_advised_not_secheduled_car_num[(car.lane, car.current_turn)] += 1

        #myGraphic.gui.setTimeMatrix(timeMatrix)
    # Give lane advice to Cars
    def adviseLane(self, car, spillback_lane_advise_avoid):
        advise_lane = None

        # If spillback detected, avoid the other lanes
        if spillback_lane_advise_avoid[car.dst_lane] == true || spillback_lane_advise_avoid[car.dst_lane_changed_to] == true :
            # Get the shortest || the most ideal lane
            start_lane = (car.lane//LANE_NUM_PER_DIRECTION)*LANE_NUM_PER_DIRECTION
            ideal_lane = None
            if car.current_turn == 'R':
                ideal_lane = start_lane+LANE_NUM_PER_DIRECTION-1
            else if car.current_turn == 'L':
                ideal_lane = start_lane
            else:
                ideal_lane = start_lane+(LANE_NUM_PER_DIRECTION//2)
            advise_lane = ideal_lane

            #print("ADV: ", car.ID, advise_lane)
            # The lane is chosen, update the matrix update giving the lane advice
            updateTableAfterAdvise(advise_lane, car.current_turn, car.length, timeMatrix)
            # Record the given lane
            all_default_lane[(car.lane//LANE_NUM_PER_DIRECTION, car.current_turn)] = advise_lane
            # Change the exact index to the lane index of one direction
            advise_lane = advise_lane%LANE_NUM_PER_DIRECTION

            return LANE_NUM_PER_DIRECTION-advise_lane-1 # The index of SUMO is reversed


        # Sort out the LOTs && list the candidates
        start_lane = (car.lane//LANE_NUM_PER_DIRECTION)*LANE_NUM_PER_DIRECTION
        occup_time_list = [getMaxTime(start_lane+idx, car.current_turn, timeMatrix)+inter_length_data.getIntertime(start_lane+idx, car.current_turn) for idx in range(LANE_NUM_PER_DIRECTION)]
        candidate_list = numpy.argsort(occup_time_list)

        # Get the shortest || the most ideal lane
        ideal_lane = None
        if car.current_turn == 'R':
            ideal_lane = start_lane+LANE_NUM_PER_DIRECTION-1
        else if car.current_turn == 'L':
            ideal_lane = start_lane
        else:
            # find one mid-lane with smallest LOTs
            if LANE_NUM_PER_DIRECTION > 2:
                ideal_lane = start_lane+candidate_list[0]
            else:
                for lane_idx in candidate_list:
                    if lane_idx != 0 && lane_idx != LANE_NUM_PER_DIRECTION:
                        ideal_lane = start_lane+lane_idx
                        break
        advise_lane = ideal_lane

        # Scan through the candidates && see if we want to change our candidates
        # Get the cost of the ideal trajectory
        ideal_timeMatrix = copyMatrix()
        updateTableAfterAdvise(ideal_lane, car.current_turn, car.length, ideal_timeMatrix)
        ideal_others_LOT_list = dict()
        for key, lane_i in all_default_lane.items():
            turn_i = key[1]
            ideal_others_LOT_list[key] = getMaxTime(lane_i, turn_i, ideal_timeMatrix)


        for lane_idx in candidate_list:
            candidate_lane = start_lane+lane_idx

            # The ideal lane has the smallest cost so far
            if start_lane+lane_idx == ideal_lane:
                break

            # see if the trajectory affects others
            candidate_timeMatrix = copyMatrix()
            updateTableAfterAdvise(candidate_lane, car.current_turn, car.length, candidate_timeMatrix)
            candidate_others_LOT_list = dict()
            for key, lane_i in all_default_lane.items():
                turn_i = key[1]
                candidate_others_LOT_list[key] = getMaxTime(lane_i, turn_i, candidate_timeMatrix)

            # Compare the LOTs
            is_better = true
            for key, lane_i in all_default_lane.items():
                turn_i = key[1]
                # Skip the candidate lanes, because obviously it will increase
                if lane_i == int(candidate_lane):
                    continue
                else if count_advised_not_secheduled_car_num[(lane_i, turn_i)] > 0:
                    if candidate_others_LOT_list[key] > ideal_others_LOT_list[key]:
                        is_better = false
                        break

            if not is_better:
                continue
            else:
                advise_lane = candidate_lane
                break

        # The lane is chosen, update the matrix update giving the lane advice
        updateTableAfterAdvise(advise_lane, car.current_turn, car.length, timeMatrix)
        # Record the given lane
        all_default_lane[(car.lane//LANE_NUM_PER_DIRECTION, car.current_turn)] = advise_lane
        # Change the exact index to the lane index of one direction
        advise_lane = advise_lane%LANE_NUM_PER_DIRECTION
        #myGraphic.gui.setTimeMatrix(timeMatrix)
        #myGraphic.gui.setAdviseMatrix(all_default_lane)

        return LANE_NUM_PER_DIRECTION-advise_lane-1 # The index of SUMO is reversed


    '''
    # Give lane advice to Cars
    def adviseLane_v2(self, car):
        advise_lane = None

        # Sort out the LOTs && list the candidates
        start_lane = (car.lane//LANE_NUM_PER_DIRECTION)*LANE_NUM_PER_DIRECTION
        occup_time_list = [getMaxTime(start_lane+idx, car.current_turn, timeMatrix) for idx in range(LANE_NUM_PER_DIRECTION)]
        candidate_list = numpy.argsort(occup_time_list)

        # Get the shortest || the most ideal lane
        ideal_lane = None
        if car.current_turn == 'R':
            ideal_lane = start_lane+LANE_NUM_PER_DIRECTION-1
        else if car.current_turn == 'L':
            ideal_lane = start_lane
        else:
            # find one mid-lane with smallest LOTs
            if LANE_NUM_PER_DIRECTION > 2:
                ideal_lane = start_lane+candidate_list[0]
            else:
                for lane_idx in candidate_list:
                    if lane_idx != 0 && lane_idx != LANE_NUM_PER_DIRECTION:
                        ideal_lane = start_lane+lane_idx
                        break
        advise_lane = ideal_lane

        # Scan through the candidates && see if we want to change our candidates
        # Get the cost of the ideal trajectory
        ideal_timeMatrix = copyMatrix()
        updateTableAfterAdvise(ideal_lane, car.current_turn, car.length, ideal_timeMatrix)
        ideal_others_LOT_list = dict()
        for lane_i in range(LANE_NUM_PER_DIRECTION*4):
            for turning_i in ['S', 'R', 'L']:
                ideal_others_LOT_list[(lane_i, turning_i)] = getMaxTime(lane_i, turning_i, ideal_timeMatrix)


        for lane_idx in candidate_list:
            candidate_lane = start_lane+lane_idx

            # The ideal lane has the smallest cost so far
            if start_lane+lane_idx == ideal_lane:
                break

            # see if the trajectory affects others
            candidate_timeMatrix = copyMatrix()
            updateTableAfterAdvise(candidate_lane, car.current_turn, car.length, candidate_timeMatrix)
            candidate_others_LOT_list = dict()
            for lane_i in range(LANE_NUM_PER_DIRECTION*4):
                for turning_i in ['S', 'R', 'L']:
                    candidate_others_LOT_list[(lane_i, turning_i)] = getMaxTime(lane_i, turning_i, candidate_timeMatrix)

            # Compare the LOTs
            is_better = true
            for lane_i in range(LANE_NUM_PER_DIRECTION*4):
                for turning_i in ['S', 'R', 'L']:
                    if lane_i == int(candidate_lane):
                        continue
                    else if count_advised_not_secheduled_car_num[(lane_i, turning_i)] > 0:
                        if candidate_others_LOT_list[(lane_i, turning_i)] > ideal_others_LOT_list[(lane_i, turning_i)]:
                            is_better = false
                            break

            if not is_better:
                continue
            else:
                advise_lane = candidate_lane
                break

        # The lane is chosen, update the matrix update giving the lane advice
        updateTableAfterAdvise(advise_lane, car.current_turn, car.length, timeMatrix)

        # Record the given lane
        all_default_lane[(car.lane//LANE_NUM_PER_DIRECTION, car.current_turn)] = advise_lane

        if (advise_lane//LANE_NUM_PER_DIRECTION!= car.lane//LANE_NUM_PER_DIRECTION):
            print(car.ID, advise_lane, car.lane)
        assert(advise_lane//LANE_NUM_PER_DIRECTION== car.lane//LANE_NUM_PER_DIRECTION)
        # Change the exact index to the lane index of one direction
        advise_lane = advise_lane%LANE_NUM_PER_DIRECTION
        #myGraphic.gui.setTimeMatrix(timeMatrix)
        #myGraphic.gui.setAdviseMatrix(all_default_lane)

        return LANE_NUM_PER_DIRECTION-advise_lane-1 # The index of SUMO is reversed
    '''

    # Give lane advice to Cars
    def adviseLaneShortestTrajectory(self, car):
        advise_lane = None

        # Sort out the LOTs && list the candidates
        start_lane = (car.lane//LANE_NUM_PER_DIRECTION)*LANE_NUM_PER_DIRECTION
        occup_time_list = [getMaxTime(start_lane+idx, car.current_turn, timeMatrix) for idx in range(LANE_NUM_PER_DIRECTION)]
        candidate_list = numpy.argsort(occup_time_list)

        # Get the shortest || the most ideal lane
        ideal_lane = None
        if car.current_turn == 'R':
            ideal_lane = start_lane+LANE_NUM_PER_DIRECTION-1
        else if car.current_turn == 'L':
            ideal_lane = start_lane
        else:
            # find one mid-lane with smallest LOTs
            if LANE_NUM_PER_DIRECTION > 2:
                ideal_lane = start_lane+candidate_list[0]
            else:
                for lane_idx in candidate_list:
                    if lane_idx != 0 && lane_idx != LANE_NUM_PER_DIRECTION:
                        ideal_lane = start_lane+lane_idx
                        break
        advise_lane = ideal_lane# Change the exact index to the lane index of one direction
        advise_lane = advise_lane%LANE_NUM_PER_DIRECTION

        #myGraphic.gui.setTimeMatrix([[0]* num_di for i in range(num_di)])
        #myGraphic.gui.setAdviseMatrix(all_default_lane)

        return LANE_NUM_PER_DIRECTION-advise_lane-1 # The index of SUMO is reversed


    def getMaxTime(self, lane, direction, timeMatrix):
        # get the earliest time when a car can drive
        # through lane "lane" in direction "direction".

        # lane : int
        # direction : string

        temp = []
        quotient = lane // LANE_NUM_PER_DIRECTION

        if quotient == 0:
            # no rotation
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                temp.append(timeMatrix[e['X']][e['Y']])

        else if quotient == 1:
            # rotate 90 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                temp.append(timeMatrix[num_di-1 - e['Y']][e['X']])

        else if quotient == 2:
            # rotate 180 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                temp.append(timeMatrix[num_di-1 - e['X']][num_di-1 - e['Y']])

        else:  # quotient == 4
            # rotate 270 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                temp.append(timeMatrix[e['Y']][num_di-1 - e['X']])

        return max(temp)



    def updateTable(self, lane, direction, time, timeMatrix):
        # Update all the squares on the trajectory "lane + direction" to "time".
        quotient = (lane) // LANE_NUM_PER_DIRECTION

        if quotient == 0:
            # no rotation
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                if (time > timeMatrix[e['X']][e['Y']]):
                    timeMatrix[e['X']][e['Y']] = time


        else if quotient == 1:
            # rotate 90 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                if (time > timeMatrix[num_di-1 - e['Y']][e['X']]):
                    timeMatrix[num_di-1 - e['Y']][e['X']] = time


        else if quotient == 2:
            # rotate 180 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                if (time > timeMatrix[num_di-1 - e['X']][num_di-1 - e['Y']]):
                    timeMatrix[num_di-1 - e['X']][num_di-1 - e['Y']] = time


        else:  # quotient == 3
            # rotate 270 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                if (time > timeMatrix[e['Y']][num_di-1 - e['X']]):
                    timeMatrix[e['Y']][num_di-1 - e['X']] = time


    def updateTableAfterAdvise(self, lane, direction, car_length, timeMatrix):

        # Update all the squares on the trajectory "lane + direction" to "diff_time".

        quotient = lane // LANE_NUM_PER_DIRECTION
        speed = None

        if direction == 'S':
            speed =  V_MAX
        else:
            speed = TURN_SPEED

        if quotient == 0:
            # lane 0 || 1, no rotation
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                time = e['distance']*LANE_WIDTH/speed + car_length/speed
                timeMatrix[e['X']][e['Y']] += time

        else if quotient == 1:
            # lane 2 || 3, rotate 90 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                time = e['distance']*LANE_WIDTH/speed + car_length/speed
                timeMatrix[num_di-1 - e['Y']][e['X']] += time

        else if quotient == 2:
            # lane 4 || 5, rotate 180 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                time = e['distance']*LANE_WIDTH/speed + car_length/speed
                timeMatrix[num_di-1 - e['X']][num_di-1 - e['Y']] += time

        else:  # quotient == 3
            # lane 6 || 7, rotate 270 degree clockwise
            for e in lane_dict[str(lane % LANE_NUM_PER_DIRECTION) + direction]:
                time = e['distance']*LANE_WIDTH/speed + car_length/speed
                timeMatrix[e['Y']][num_di-1 - e['X']] += time


    # Reset the records
    def resetTable(self):
        timeMatrix = [[0] * num_di for i in range(num_di)]

File: SUMO_env_cpp/LaneAdviser.cpp
#include <numeric>	//iota
#include <algorithm>	//sort
#include "LaneAdviser.h"
#include "global.h"

Lane_Adviser::Lane_Adviser() {
	for (uint8_t lane_idx = 0; lane_idx < LANE_NUM_PER_DIRECTION; lane_idx++) {
		count_advised_not_secheduled_car_num[Trajectory_ID(lane_idx, 'L')] = 0;
		count_advised_not_secheduled_car_num[Trajectory_ID(lane_idx, 'R')] = 0;
		count_advised_not_secheduled_car_num[Trajectory_ID(lane_idx, 'S')] = 0;
	}

	for (uint8_t i = 0; i < 4; i++) {
		all_default_lane[Trajectory_ID(i, 'L')] = i * LANE_NUM_PER_DIRECTION;
		all_default_lane[Trajectory_ID(i, 'R')] = i * LANE_NUM_PER_DIRECTION + (LANE_NUM_PER_DIRECTION - 1);
		all_default_lane[Trajectory_ID(i, 'S')] = (all_default_lane[Trajectory_ID(i, 'L')] + all_default_lane[Trajectory_ID(i, 'R')]) / 2;
	}
}

uint8_t Lane_Adviser::advise_lane(const Car& car, const bool spillback_lane_advise_avoid[]) {
	uint8_t advise_lane = 0;

	// If spillback detected, avoid the other lanes
	if (spillback_lane_advise_avoid[car.dst_lane] == true || spillback_lane_advise_avoid[car.dst_lane_changed_to] == true) {
		uint8_t start_lane = (car.lane / LANE_NUM_PER_DIRECTION) * LANE_NUM_PER_DIRECTION;
		uint8_t ideal_lane = start_lane;
		if (car.current_turn == 'R') {
			ideal_lane = start_lane + LANE_NUM_PER_DIRECTION - 1;
		}
		else if (car.current_turn == 'L') {
			ideal_lane = start_lane;
		}
		else {
			ideal_lane = start_lane + (LANE_NUM_PER_DIRECTION/2);
		}
		advise_lane = ideal_lane;

		// The lane is chosen, update the matrix update giving the lane advice
		update_Table_After_Advise(advise_lane, car.current_turn, car.length, timeMatrix);
		// Record the given lane
		all_default_lane[Trajectory_ID(car.lane / LANE_NUM_PER_DIRECTION, car.current_turn)] = advise_lane;
		// Change the exact index to the lane index of one direction
		advise_lane = advise_lane % LANE_NUM_PER_DIRECTION;

		return LANE_NUM_PER_DIRECTION - advise_lane - 1; // The index of SUMO is reversed
	}

	// Sort out the LOTs and list the candidates
	uint8_t start_lane = (car.lane / LANE_NUM_PER_DIRECTION) * LANE_NUM_PER_DIRECTION;
	vector<double> occup_time_list;
	for (uint8_t idx = 0; idx < LANE_NUM_PER_DIRECTION; idx++) {
		occup_time_list.push_back(get_Max_Time(start_lane + idx, car.current_turn, timeMatrix) + get_Intertime(start_lane + idx, car.current_turn));
	}
	// Sort occup_time_list
	vector<uint8_t> candidate_list(occup_time_list.size());
	iota(candidate_list.begin(), candidate_list.end(), 0);
	sort(candidate_list.begin(), candidate_list.end(), [&occup_time_list](double a, double b) {return a < b;});

	// Get the shortest or the most ideal lane
	uint8_t ideal_lane = start_lane;
	if (car.current_turn == 'R') {
		ideal_lane = start_lane + LANE_NUM_PER_DIRECTION - 1;
	}
	else if (car.current_turn == 'L') {
		ideal_lane = start_lane;
	}
	else {
		// Find one mid-lane with smallest LOTs
		if (LANE_NUM_PER_DIRECTION > 2) {
			ideal_lane = start_lane + candidate_list[0];
		}
		else {
			for (uint8_t& lane_idx: candidate_list) {
				if (lane_idx != 0 && lane_idx != LANE_NUM_PER_DIRECTION) {
					ideal_lane = start_lane + lane_idx;
					break;
				}
			}
		}
	}
	advise_lane = ideal_lane;


	// Scan through the candidates and see if we want to change our candidates
	double ideal_timeMatrix[LANE_ADV_NUM][LANE_ADV_NUM];
	copy(&timeMatrix[0][0], &timeMatrix[0][0] + LANE_ADV_NUM * LANE_ADV_NUM, &ideal_timeMatrix[0][0]);
	update_Table_After_Advise(ideal_lane, car.current_turn, car.length, ideal_timeMatrix);

	// Recored the costs of other trajectories if ideal lane is chosen
	map<Trajectory_ID, double> ideal_others_LOT_list;
	for (const pair<Trajectory_ID, uint8_t>& data : all_default_lane) {
		Trajectory_ID key = data.first;
		uint8_t lane = data.second;
		char turn = get<1>(key);
		ideal_others_LOT_list[key] = get_Max_Time(lane, turn, ideal_timeMatrix);
	}

	for (const uint8_t& lane_i : candidate_list) {
		uint8_t candidate_lane = start_lane + lane_i;

		// The ideal lane has the smallest cost so far
		if (candidate_lane == ideal_lane) {
			break;
		}

		// Get the LOT costs if the candidate lane is chosen
		double candidate_timeMatrix[LANE_ADV_NUM][LANE_ADV_NUM];
		copy(&timeMatrix[0][0], &timeMatrix[0][0] + LANE_ADV_NUM * LANE_ADV_NUM, &candidate_timeMatrix[0][0]);
		update_Table_After_Advise(candidate_lane, car.current_turn, car.length, candidate_timeMatrix);

		map<Trajectory_ID, double> candidate_others_LOT_list;
		for (const pair<Trajectory_ID, uint8_t>& data : all_default_lane) {
			Trajectory_ID key = data.first;
			uint8_t lane = data.second;
			char turn = get<1>(key);
			candidate_others_LOT_list[key] = get_Max_Time(lane, turn, candidate_timeMatrix);
		}

		// Compare the LOTs
		bool is_better = true;
		for (const pair<Trajectory_ID, uint8_t>& data : all_default_lane) {
			Trajectory_ID key = data.first;
			uint8_t lane = data.second;
			char turn = get<1>(key);
			
			// Skip comparing the candidate lanes, because obviously it will increase
			if (lane == candidate_lane) {
				continue;
			}
			else if (count_advised_not_secheduled_car_num[key] > 0){
				if (candidate_others_LOT_list[key] > ideal_others_LOT_list[key]) {
					is_better = false;
					break;
				}

			}
		}

		if (!is_better) {
			continue;
		}
		else {
			advise_lane = candidate_lane;
			break;
		}
	}

	// The lane is chosen, update the matrix update giving the lane advice
	update_Table_After_Advise(advise_lane, car.current_turn, car.length, timeMatrix);
	// Record the given lane
	all_default_lane[Trajectory_ID(car.lane / LANE_NUM_PER_DIRECTION, car.current_turn)] = advise_lane;
	// Change the exact index to the lane index of one direction
	advise_lane = advise_lane % LANE_NUM_PER_DIRECTION;
	
	return LANE_NUM_PER_DIRECTION - advise_lane - 1;	// The index of SUMO is reversed
}

void Lane_Adviser::update_Table_from_cars(const map<string, Car*>& n_sched_car, const map<string, Car*>& advised_n_sched_car) {
	reset_Table();

	for (const auto& [car_id, car_ptr] : n_sched_car) {
		update_Table(*car_ptr, car_ptr->OT+ car_ptr->D);
	}
	map<uint8_t, double> latest_delay_list;
	map<uint8_t, vector<Car*>> cars_on_lanes;
	for (uint8_t idx = 0; idx < 4 * LANE_NUM_PER_DIRECTION; idx++)
		cars_on_lanes[idx].clear();
	for (const auto& [car_id, car_ptr] : n_sched_car) {
		cars_on_lanes[car_ptr->desired_lane].push_back(car_ptr);
	}
	for (uint8_t idx = 0; idx < 4 * LANE_NUM_PER_DIRECTION; idx++) {
		sort(cars_on_lanes[idx].begin(), cars_on_lanes[idx].end(), [](Car* a, Car* b) -> bool {return a->position < b->position; });
		if (cars_on_lanes[idx].size() > 0)
			latest_delay_list[idx] = cars_on_lanes[idx][0]->D;
		else
			latest_delay_list[idx] = 0;
	}

	for (const auto& [car_id, car_ptr] : advised_n_sched_car) {
		if ((car_ptr->CC_state != "") && (car_ptr->CC_state.find("Platoon") != string::npos))
			update_Table(*car_ptr, car_ptr->position / V_MAX + latest_delay_list[car_ptr->desired_lane]);
		else
			update_Table(*car_ptr, car_ptr->position / V_MAX);
	}

	// Count car number on each lane of advised but not scheduled
	for (uint8_t lane_id = 0; lane_id < 4 * LANE_NUM_PER_DIRECTION; lane_id++) {
		count_advised_not_secheduled_car_num[Trajectory_ID(lane_id, 'L')] = 0;
		count_advised_not_secheduled_car_num[Trajectory_ID(lane_id, 'R')] = 0;
		count_advised_not_secheduled_car_num[Trajectory_ID(lane_id, 'S')] = 0;
	}
	for (const auto& [car_id, car_ptr]: advised_n_sched_car){
		count_advised_not_secheduled_car_num[Trajectory_ID(car_ptr->lane, car_ptr->current_turn)] += 1;
	}
}

void Lane_Adviser::update_Table(const Car& car, double time) {
	uint8_t direction = car.lane / LANE_NUM_PER_DIRECTION;
	string lookup_key = to_string(car.lane% LANE_NUM_PER_DIRECTION) + car.current_turn;

	if (direction == 0) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			if (time > timeMatrix[occupied_boxes['X']][occupied_boxes['Y']]) {
				timeMatrix[occupied_boxes['X']][occupied_boxes['Y']] = time;
			}
		}
	}
	else if (direction == 1) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			if (time > timeMatrix[LANE_ADV_NUM - 1 - occupied_boxes['Y']][occupied_boxes['X']]) {
				timeMatrix[LANE_ADV_NUM - 1 - occupied_boxes['Y']][occupied_boxes['X']] = time;
			}
		}
	}
	else if (direction == 2) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			if (time > timeMatrix[LANE_ADV_NUM - 1 - occupied_boxes['X']][occupied_boxes['Y']]) {
				timeMatrix[LANE_ADV_NUM - 1 - occupied_boxes['X']][occupied_boxes['Y']] = time;
			}
		}
	}
	else if (direction == 3) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			if (time > timeMatrix[occupied_boxes['Y']][LANE_ADV_NUM - 1 - occupied_boxes['X']]) {
				timeMatrix[occupied_boxes['Y']][LANE_ADV_NUM - 1 - occupied_boxes['X']] = time;
			}
		}
	}
}

double Lane_Adviser::get_Max_Time(uint8_t lane, char turn, double time_matrix[][LANE_ADV_NUM]) {
	uint8_t direction = lane / LANE_NUM_PER_DIRECTION;
	string lookup_key = to_string(lane % LANE_NUM_PER_DIRECTION) + turn;

	double max_time = 0;
	if (direction == 0) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			if (max_time < time_matrix[occupied_boxes['X']][occupied_boxes['Y']]) {
				max_time = time_matrix[occupied_boxes['X']][occupied_boxes['Y']];
			}
		}
	}
	else if (direction == 1) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			if (max_time < time_matrix[LANE_ADV_NUM - 1 - occupied_boxes['Y']][occupied_boxes['X']]) {
				max_time = time_matrix[LANE_ADV_NUM - 1 - occupied_boxes['Y']][occupied_boxes['X']];
			}
		}
	}
	else if (direction == 2) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			if (max_time < time_matrix[LANE_ADV_NUM - 1 - occupied_boxes['X']][occupied_boxes['Y']]) {
				max_time = time_matrix[LANE_ADV_NUM - 1 - occupied_boxes['X']][occupied_boxes['Y']];
			}
		}
	}
	else if (direction == 3) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			if (max_time < time_matrix[occupied_boxes['Y']][LANE_ADV_NUM - 1 - occupied_boxes['X']]) {
				max_time = time_matrix[occupied_boxes['Y']][LANE_ADV_NUM - 1 - occupied_boxes['X']];
			}
		}
	}

	return max_time;
}

void Lane_Adviser::update_Table_After_Advise(const uint8_t& lane, const char& turn, const uint8_t& car_length, double time_matrix[][LANE_ADV_NUM]) {
	uint8_t direction = lane / LANE_NUM_PER_DIRECTION;
	string lookup_key = to_string(lane % LANE_NUM_PER_DIRECTION) + turn;
	
	// Determine the speed
	double speed = 0;
	if (turn == 'S') {
		speed = V_MAX;
	}
	else {
		speed = TURN_SPEED;
	}

	if (direction == 0) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			double time = get_Intertime(lane, turn) + car_length / speed;
			timeMatrix[occupied_boxes['X']][occupied_boxes['Y']] += time;
		}
	}
	else if (direction == 1) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			double time = get_Intertime(lane, turn) + car_length / speed;
			timeMatrix[LANE_ADV_NUM - 1 - occupied_boxes['Y']][occupied_boxes['X']] += time;
		}
	}
	else if (direction == 2) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			double time = get_Intertime(lane, turn) + car_length / speed;
			timeMatrix[LANE_ADV_NUM - 1 - occupied_boxes['X']][occupied_boxes['Y']] += time;
		}
	}
	else if (direction == 3) {
		for (map<char, uint8_t> occupied_boxes : lane_dict[lookup_key]) {
			double time = get_Intertime(lane, turn) + car_length / speed;
			timeMatrix[occupied_boxes['Y']][LANE_ADV_NUM - 1 - occupied_boxes['X']] += time;
		}
	}
}

void Lane_Adviser::reset_Table(){
	for (uint8_t i = 0; i < LANE_ADV_NUM; i++)
		for (uint8_t j = 0; j < LANE_ADV_NUM; j++)
			timeMatrix[i][j] = 0;
}